File: ios/Classes/ApplePay/ApplePayHandler.m
#import "ApplePayHandler.h"

@implementation ApplePayHandler {
    FlutterResult flutterResult;
    ApplePayRegistrarChannel* channel;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        channel = [ApplePayRegistrarChannel new];
    }
    return self;
}

- (void) start: (NSDictionary *) dictionary result:(FlutterResult) result {
    flutterResult = result;
    UIViewController * viewController = [RootViewControllerProvider get];

    bool isSandbox = [[dictionary objectForKey:@"isSandbox"] boolValue];

    P24ApplePayParams* params = [self parseApplePayParams: dictionary];

    params.sandbox = isSandbox;
    [P24 startApplePay:params inViewController: viewController delegate:self];
}

- (P24ApplePayParams *) parseApplePayParams: (NSDictionary *) dictionary {
    NSString* appleMerchantId = [dictionary valueForKey:@"appleMerchantId"];
    NSString* currency = [dictionary objectForKey:@"currency"];
    NSArray<PaymentItem *> * items = [self getItemsList: dictionary];

    return [[P24ApplePayParams alloc]
            initWithItems: items
            currency: currency
            appleMerchantId: appleMerchantId
            registrar: channel];
}

- (NSArray<PaymentItem *> *) getItemsList: (NSDictionary *) dictionary {
    NSArray<NSDictionary *> * itemsDictionary = [dictionary valueForKey:@"items"];
    NSMutableArray<PaymentItem *> * paymentItemsList = [[NSMutableArray<PaymentItem *> alloc] init];

    for(NSDictionary * singleItem in itemsDictionary){
        PaymentItem * item = [[PaymentItem alloc] init];
        item.amount = [[singleItem objectForKey:@"amount"] intValue];
        item.itemDescription = [singleItem objectForKey:@"description"];
        [paymentItemsList addObject: item];
    }

    return paymentItemsList;
}

- (void)p24ApplePayOnCanceled {
    flutterResult(@{ @"status" : @"cancel"});
}

- (void)p24ApplePayOnError:(NSString *)errorCode {
    flutterResult(@{ @"status" : @"error", @"payload" : errorCode});
}

- (void)p24ApplePayOnSuccess {
    flutterResult(@{ @"status" : @"success"});
}


@end

File: ios/Classes/ApplePay/ApplePayHandler.swift
import P24

internal class ApplePayHandler: NSObject {
    
    private let flutterResult: FlutterResult
    private let registar: ApplePayRegistrarChannel
    
    internal init(flutterResult: @escaping FlutterResult, registar: ApplePayRegistrarChannel) {
        self.flutterResult = flutterResult
        self.registar = registar
    }
    
    internal func start(dictionary: [String: Any]) {
        let params = parseApplePayParams(dictionary)
        params.sandbox = dictionary["isSandbox"] as! Bool
        P24.startApplePay(params, in: RootViewControllerProvider.shared, delegate: self)
    }
    
    private func parseApplePayParams(_ dict: [String: Any]) -> P24ApplePayParams {
        let appleMerchantId = dict["appleMerchantId"] as! String
        let currency = dict["currency"] as! String
        let items = getItemsList(dict["items"] as! [[String: Any]])
        return P24ApplePayParams(items: items, currency: currency, appleMerchantId: appleMerchantId, registrar: registar)
    }
    
    private func getItemsList(_ itemsList: [[String: Any]]) -> [PaymentItem] {
        return itemsList.map {
            let item = PaymentItem()
            item.amount = $0["amount"] as! Int32
            item.itemDescription = ($0["description"] as! String)
            return item
        }
    }
}

extension ApplePayHandler: P24ApplePayDelegate {
    func p24ApplePayOnSuccess() {
        flutterResult(["status" : "success"])
    }
    
    func p24ApplePay(onError errorCode: String!) {
        flutterResult(["status" : "error", "payload" : errorCode])
    }
    
    func p24ApplePayOnCanceled() {
        flutterResult(["status" : "cancel"])
    }
    
}
